### PR TITLE
[CheriBoundAllocas] Bug fixes and enable one-deep analysis for -O0/optnone

### PR DESCRIPTION
**Pull request by jrtc27**

### CODE REVIEW

File: llvm/lib/Analysis/CheriBounds.cpp [hunk 1]
```diff
@@ -112,7 +112,8 @@ bool CheriNeedBoundsChecker::useNeedsBounds(const Use &U,
           APInt Zero(CurrentGEPOffset.getBitWidth(), 0);
           APInt Max(CurrentGEPOffset.getBitWidth(), *MinSizeInBytes);
           APInt LastAddr = CurrentGEPOffset + CI->getValue();
```
[the remark this answers is not in the document]
Comment (arichardson) [reply llvm/lib/Analysis/CheriBounds.cpp:119]: Looks like there is, could you change this to use `uadd_ov()`?

File: llvm/lib/Analysis/CheriBounds.cpp [hunk 2]
```diff
@@ -299,7 +306,8 @@ bool CheriNeedBoundsChecker::canLoadStoreBeOutOfBounds(
   APInt Zero(CurrentGEPOffset.getBitWidth(), 0);
   APInt Max(CurrentGEPOffset.getBitWidth(), *MinSizeInBytes);
   APInt LastAddr = CurrentGEPOffset + Size;
-  if (CurrentGEPOffset.slt(Zero) || LastAddr.sgt(Max)) {
+  if (CurrentGEPOffset.slt(Zero) || LastAddr.slt(CurrentGEPOffset) ||
```
Comment (arichardson) [inline llvm/lib/Analysis/CheriBounds.cpp:309]: Same here, `uadd_ov`

File: llvm/lib/Analysis/CheriBounds.cpp [hunk 3]
```diff
@@ -13,14 +13,18 @@ using namespace llvm;
...
 // This is similar to CaptureTracking but we treat way more cases as "captured"
-bool CheriNeedBoundsChecker::check(const Use &U) const {
+bool CheriNeedBoundsChecker::check(const Use &U, bool Simple) const {
   if (!MinSizeInBytes) {
     DBG_MESSAGE("dyanmic size alloca always needs bounds: ";
                 U.getUser()->dump());
     return true;
   }
   APInt CurrentGEPOffset = APInt(DL.getIndexSizeInBits(PointerAS), 0);
-  bool Result = useNeedsBounds(U, CurrentGEPOffset, 1);
```
Comment (arichardson) [inline llvm/lib/Analysis/CheriBounds.cpp:23]: Could you add a test for this change? For example a function with three allocas, load, store, gep?

**Review by arichardson — COMMENTED** [review]
Thanks, that should massively improve O0 codegen. Just a few small review comments.

**Comment by arichardson** [comment]
Sigh more LLD changes written in C... Sorry about that :(

File: llvm/lib/Analysis/CheriBounds.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jrtc27) [reply llvm/lib/Analysis/CheriBounds.cpp:119]: I assume you mean sadd_ov?

File: llvm/lib/CodeGen/CheriBoundAllocas.cpp [hunk 4]
```diff
@@ -66,10 +65,25 @@ static cl::opt<StackBoundsMethod> BoundsSettingMode(
...
+    "cheri-stack-bounds-analysis",
+    cl::desc("Strategy for analysing bounds for stack capabilities:"),
+    cl::init(StackBoundsAnalysis::Default),
+    cl::values(clEnumValN(StackBoundsAnalysis::Default, "default",
+                          "Use the default strategy (none for "
+                          "-O0/optnone, full otherwise)"),
+               clEnumValN(StackBoundsAnalysis::None, "none",
+                          "Assume all uses require bounds"),
+               clEnumValN(StackBoundsAnalysis::Full, "full",
+                          "Fully analyse whether bounds are required")));
```
[the remark this answers is not in the document]
Comment (jrtc27) [reply llvm/lib/CodeGen/CheriBoundAllocas.cpp:89]: 10 is indeed smaller than infinity, plus we wouldn't catch everything even with an unbounded depth... hm

File: llvm/lib/Analysis/CheriBounds.cpp (hunk 3 above)
Comment (jrtc27) [reply llvm/lib/Analysis/CheriBounds.cpp:23]: As in an explicit optnone test that illustrates bounds being omitted for load+store but not GEP+load/store?

File: llvm/lib/Analysis/CheriBounds.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (arichardson) [reply llvm/lib/Analysis/CheriBounds.cpp:119]: The CurrentGEPOffset is an unsigned APInt, so I think we should be using uadd_ov?

[the remark this answers is not in the document]
Comment (arichardson) [reply llvm/lib/Analysis/CheriBounds.cpp:119]: It looks like this code is somewhat confused about signed vs unsigned...

[the remark this answers is not in the document]
Comment (jrtc27) [reply llvm/lib/Analysis/CheriBounds.cpp:119]: APInts don't track signedness, they're like IR in that their signedness is based on how you use them

[the remark this answers is not in the document]
Comment (jrtc27) [reply llvm/lib/Analysis/CheriBounds.cpp:119]: (and GEPOperator::accumulateConstantOffset will correctly treat them as signed like LangRef says they should be and like we do for comparisons)

[the remark this answers is not in the document]
Comment (jrtc27) [reply llvm/lib/Analysis/CheriBounds.cpp:119]: LLVM doesn't support allocations bigger than IndexWidth/2

[the remark this answers is not in the document]
Comment (jrtc27) [reply llvm/lib/Analysis/CheriBounds.cpp:119]: (precisely because of signedness reasons)

[the remark this answers is not in the document]
Comment (jrtc27) [reply llvm/lib/Analysis/CheriBounds.cpp:119]: > These rules are based on the assumption that no allocated object may cross the unsigned address space boundary, and no allocated object may be larger than half the pointer index type space.

From LangRef's description of the inbounds GEP flag

[the remark this answers is not in the document]
Comment (arichardson) [reply llvm/lib/Analysis/CheriBounds.cpp:119]: Perfect thanks for looking that up, in that case just sadd_ov 🙂

**Review by arichardson — APPROVED** [review]
LGTM. Clang-format isn't happy so I'd suggest fixing that unless you feel it's much worse than the current style.

**Comment by jrtc27** [comment]
> LGTM. Clang-format isn't happy so I'd suggest fixing that unless you feel it's much worse than the current style.

Eh, some of it is reformatting existing code, other parts are where it decides on one option and I chose another valid one...

**Comment by arichardson** [comment]
> Eh, some of it is reformatting existing code, other parts are where it decides on one option and I chose another valid one...

Yeah I don't care which one is used, so let's go ahead and merge this :)

**Pull request merged**